Cython/Compiler/Code.py

```diff
@@ -2685,6 +2686,30 @@ def put_pymethoddef_wrapper(self, entry):
                 self.putln("}")
         return func_cname
 
+    def put_pyobject_lock(self, cname):
+        self.globalstate.use_utility_code(
+            UtilityCode.load_cached("AccessCriticalSectionForFreeThreading", "ModuleSetupCode.c"))
+        self.putln("#if CYTHON_COMPILING_IN_CPYTHON_FREETHREADING")
+        self.putln(f"Py_BEGIN_CRITICAL_SECTION({cname});")
+        self.putln("#endif")
+
+    def put_pyobject_unlock(self, cname):
+        self.putln("#if CYTHON_COMPILING_IN_CPYTHON_FREETHREADING")
+        self.putln(f"Py_END_CRITICAL_SECTION({cname});")
+        self.putln("#endif")
+
+    def put_scope_pymutex_lock(self, cname, is_local):
+        if is_local:
+            self.funcstate.uses_scope_mutex = True
+        self.putln("#if CYTHON_COMPILING_IN_CPYTHON_FREETHREADING")
+        self.putln(f"PyMutex_Lock(&{cname});")
+        self.putln("#endif")
+
+    def put_scope_pymutex_unlock(self, cname):
+        self.putln("#if CYTHON_COMPILING_IN_CPYTHON_FREETHREADING")
+        self.putln(f"PyMutex_Unlock(&{cname});")
+        self.putln("#endif")
```

Comment:
I haven't looked more deeply into the implementation yet, but we can probably reduce the C code overhead here by defining macros for these, just doing nothing for other targets than `FREETHREADING`.